util/stop: TestStopperCloserConcurrent failed

[(util/stop).TestStopperCloserConcurrent failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2612746&tab=buildLog) on [master@48fbbe7313fc70f889fa66c5749914f57ddd5a91](https://github.com/cockroachdb/cockroach/commits/48fbbe7313fc70f889fa66c5749914f57ddd5a91):

```
=== RUN   TestStopperCloserConcurrent
I210125 08:44:42.607473 13 util/stop/stopper.go:561  quiescing
I210125 08:44:42.607557 15 util/stop/stopper.go:561  quiescing
I210125 08:44:42.607611 98 util/stop/stopper.go:561  quiescing
I210125 08:44:42.607654 100 util/stop/stopper.go:561  quiescing
    stopper_test.go:264: expected true; got false
I210125 08:44:42.607736 40 util/stop/stopper.go:561  quiescing
I210125 08:44:42.607839 42 util/stop/stopper.go:561  quiescing
I210125 08:44:42.607873 44 util/stop/stopper.go:561  quiescing
I210125 08:44:42.607909 46 util/stop/stopper.go:561  quiescing
I210125 08:44:42.607944 48 util/stop/stopper.go:561  quiescing
I210125 08:44:42.607989 114 util/stop/stopper.go:561  quiescing
--- FAIL: TestStopperCloserConcurrent (0.00s)

```

<details><summary>More</summary><p>
Parameters:

- TAGS=
- GOFLAGS=-parallel=4

```
make stressrace TESTS=TestStopperCloserConcurrent PKG=./pkg/util/stop TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestStopperCloserConcurrent.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(util/stop).TestStopperCloserConcurrent failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2613389&tab=buildLog) on [master@b51d9f1271a6a8277a24003ca32623fbdfcd591a](https://github.com/cockroachdb/cockroach/commits/b51d9f1271a6a8277a24003ca32623fbdfcd591a):

```
=== RUN   TestStopperCloserConcurrent
I210125 17:27:38.599387 52 util/stop/stopper.go:561  quiescing
I210125 17:27:38.599823 54 util/stop/stopper.go:561  quiescing
I210125 17:27:38.600372 56 util/stop/stopper.go:561  quiescing
I210125 17:27:38.600764 58 util/stop/stopper.go:561  quiescing
I210125 17:27:38.601097 60 util/stop/stopper.go:561  quiescing
I210125 17:27:38.601465 62 util/stop/stopper.go:561  quiescing
I210125 17:27:38.602126 15 util/stop/stopper.go:561  quiescing
I210125 17:27:38.602727 82 util/stop/stopper.go:561  quiescing
    stopper_test.go:264: expected true; got false
I210125 17:27:38.603305 84 util/stop/stopper.go:561  quiescing
I210125 17:27:38.603928 86 util/stop/stopper.go:561  quiescing
--- FAIL: TestStopperCloserConcurrent (0.01s)

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestStopperCloserConcurrent PKG=./pkg/util/stop TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestStopperCloserConcurrent.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/util/stop/stopper.go
+++ b/pkg/util/stop/stopper.go
@@ -247,12 +247,11 @@ func (s *Stopper) Recover(ctx context.Context) {
 func (s *Stopper) AddCloser(c Closer) {
 	s.mu.Lock()
 	defer s.mu.Unlock()
-	select {
-	case <-s.stopped:
+	if s.mu.stopCalled {
 		c.Close()
-	default:
-		s.mu.closers = append(s.mu.closers, c)
+		return
 	}
+	s.mu.closers = append(s.mu.closers, c)
 }
 
 // WithCancelOnQuiesce returns a child context which is canceled when the

